src/PantheonShell.vala

```diff
@@ -0,0 +1,272 @@
+ * Copyright 2023 Corentin Noël <tintou@noel.tf>
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+namespace Gala {
+#if !HAS_MUTTER45
+    [Compact]
+    public class FakeMetaWaylandCompositor : GLib.Object {
+        // It is the third field and Vala adds a FakeMetaWaylandCompositorPrivate *priv
+        public Wl.Display wayland_display;
+
+        [CCode (cname = "meta_context_get_wayland_compositor")]
+        public extern static unowned Gala.FakeMetaWaylandCompositor from_context (Meta.Context context);
+    }
+#endif
+    public static inline unowned Wl.Display? get_display_from_context (Meta.Context context) {
+#if HAS_MUTTER45
+        unowned Meta.WaylandCompositor? compositor = context.get_wayland_compositor ();
+        if (compositor == null) {
+            return null;
+        }
+
+        return (Wl.Display) compositor.get_wayland_display ();
+#else
+        unowned FakeMetaWaylandCompositor compositor = Gala.FakeMetaWaylandCompositor.from_context (context);
+        if (compositor == null) {
+            return null;
+        }
+
+        return compositor.wayland_display;
+#endif
+    }
+
+    private static Pantheon.Desktop.ShellInterface wayland_pantheon_shell_interface;
+    private static Pantheon.Desktop.PanelInterface wayland_pantheon_panel_interface;
+    private static Pantheon.Desktop.WidgetInterface wayland_pantheon_widget_interface;
+    private static Pantheon.Desktop.ExtendedBehaviorInterface wayland_pantheon_extended_behavior_interface;
+    private static Wl.Global shell_global;
+
+    public void init_pantheon_shell (Meta.Context context) {
+        unowned Wl.Display? wl_disp = get_display_from_context (context);
+        if (wl_disp == null) {
+            debug ("Not running under Wayland, no Pantheon Shell protocol");
+            return;
+        }
+
+        wayland_pantheon_shell_interface = {
+            get_panel,
+            get_widget,
+            get_extended_behavior,
+        };
+
+        wayland_pantheon_panel_interface = {
+            destroy_panel_surface,
+            set_anchor,
+        };
+
+        wayland_pantheon_widget_interface = {
+            destroy_widget_surface,
+        };
+
+        wayland_pantheon_extended_behavior_interface = {
+            destroy_extended_behavior_surface,
+            set_keep_above,
+        };
+
+        PanelSurface.quark = GLib.Quark.from_string ("-gala-wayland-panel-surface-data");
+
+        shell_global = Wl.Global.create (wl_disp, ref Pantheon.Desktop.ShellInterface.iface, 1, (client, version, id) => {
+            unowned var resource = client.create_resource (ref Pantheon.Desktop.ShellInterface.iface, (int) version, id);
+            resource.set_implementation (&wayland_pantheon_shell_interface, null, (res) => {});
+        });
+    }
+
+    public class PanelSurface : GLib.Object {
+        public static GLib.Quark quark = 0;
+        public unowned GLib.Object? wayland_surface;
+        public unowned Wl.Resource resource;
```

Comment:
Not sure whether we are going to use them at one point but I don't think we will?
```suggestion
```

Reply:
Indeed!